### PR TITLE
Implement the OTA PAL layer based on TF-M firmware update service.

### PR DESCRIPTION
**Pull request by Sherryzhang2**
Signed-off-by: Sherry Zhang <sherry.zhang2@arm.com>
Change-Id: I86a8f725c99bcd4547b7306c1401aac723a0b3be

TF-M Firmware Update(FWU) service-based OTA PAL layer

TF-M FWU service provides the functionality of updating firmware images. This PR adds the implementation of the TF-M FWU serviced based OTA PAL layer in which the FWU APIs are called. This PAL layer implementation is platform-independent.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: vendors/arm/TF-M/ports/ota/aws_ota_pal.c [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+
+#define PSA_CODE_SIGNING_KEY_ID     ( ( psa_key_id_t )0x11 )
+
+/***********************************************************************
+ *
+ * defines
+ *
+ **********************************************************************/
+#define OTA_PAL_PRINT( X )            vLoggingPrintf X
+static const char pcOTA_PAL_CERT_BEGIN[] = "-----BEGIN CERTIFICATE-----";
```
Comment (aggarg) [inline vendors/arm/TF-M/ports/ota/aws_ota_pal.c]: Unused variables can be removed.

Comment (Sherryzhang2) [reply vendors/arm/TF-M/ports/ota/aws_ota_pal.c]: Done

File: libraries/abstractions/ota_pal_psa/aws_ota_pal.c [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+ * firmware received via OTA. It is typically just a reset of the device.
+ *
+ * @note This function SHOULD not return. If it does, the platform doesn't support
+ * an automatic reset or an error occurred.
+ *
+ * @return The OTA PAL layer error code combined with the MCU specific error code. See OTA Agent
+ * error codes information in aws_iot_ota_agent.h.
+ */
+OTA_Err_t prvPAL_ActivateNewImage( void )
+{
```
Comment (pvyawaha) [inline libraries/abstractions/ota_pal_psa/aws_ota_pal.c:347]: The OTA process includes self-test phase , i.e when the the complete image is received its signature is verified and it is marked as in - Testing. The device then reboots and connects to AWS IoT services and updates the status after which image is accepted. If the device can not communicate back to the service then a timer resets the device and it should roll back to previous version. I do not see this support implemented.

File: libraries/abstractions/ota_pal_psa/aws_ota_pal.c [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+ */
+
+/* C Runtime includes. */
+#include <stdlib.h>
+#include <string.h>
+
+/* Amazon FreeRTOS include. */
+#include "FreeRTOS.h"
+#include "FreeRTOSConfig.h"
+#include "aws_iot_ota_pal.h"
```
Comment (pvyawaha) [inline libraries/abstractions/ota_pal_psa/aws_ota_pal.c:40]: Just heads up on OTA pal we have minor changes in the interface in the latest release of OTA library and pal  - https://freertos.org/Documentation/api-ref/ota-for-aws-iot-embedded-sdk/docs/doxygen/output/html/ota_pal_interface.html#ota_pal_interface_overview

File: libraries/abstractions/ota_pal_psa/aws_ota_pal.c (hunk 2 above)
Comment (Sherryzhang2) [reply libraries/abstractions/ota_pal_psa/aws_ota_pal.c:347]: After the image's signature(covering the whole .bin file) which is placed at the OTA Job when creating an OTA Job is verified when prvPAL_CloseFile is called by the OTA agent. And then if the signature verification passes, the OTA agent informs the OTA server that the image signature pass and calls the prvPAL_ActivateNewImage API to active the image. In prvPAL_ActivateNewImage, the prvPAL_ResetDevice is called to reboot the device. During the reboot, the image is validated by the bootloader in secure world(where TF-M runs). The image validation includes:

1. Signature verification. Note that this signature is different from that used when creating the OTA Job. This signature is part of the image itself and it protects the critical part of the whole image.
2. Security count check which is for rollback protection.

After reboot, in the self-test phase, it checks the image version. If the running image's version is higher than the version in the self-test, then we think the image in the OTA job is running and when the OTA agent set the image to eOTA_ImageState_Accepted state by calling prvPAL_SetPlatformImageState, the psa_fwu_accept API is called to  mark this image as a permanent one. If the image with the lower version is running, then reboot is required by the OTA agent and the original image runs again after the reboot.

File: libraries/abstractions/ota_pal_psa/aws_ota_pal.c (hunk 3 above)
Comment (Sherryzhang2) [reply libraries/abstractions/ota_pal_psa/aws_ota_pal.c:40]: Okay. I will check the update on the latest release.

Comment (Sherryzhang2) [reply libraries/abstractions/ota_pal_psa/aws_ota_pal.c:40]: I checked the latest master branch(5251eeaf01e09d24a666bec409bb45d8165a869e), the OTA pal APIs are still defined in aws_iot_ota_pal.h. I checked git log of the ./libraries/freertos_plus/aws/ota/src/aws_iot_ota_agent.c, the lastest commit was on 4 Nov 2020. So I want to double confirm with @pvyawaha , has the latest release of the OTA library you mentioned been merged into the Amazon FreeRTOS repo?

**Comment by Sherryzhang2** [comment]
Updates this PR to the latest code base(5251eeaf0) of official Amazon FreeRTOS repo.
I have tested this PR on an Arm platform with AWS OTA server. It test result shows it works well.

By the way, during the test, I found an issue with the network connection. I created the issue in https://github.com/aws/amazon-freertos/issues/3075 and fixed it locally to make the test project work.

File: libraries/abstractions/ota_pal_psa/aws_ota_pal.c (hunk 2 above)
Comment (pvyawaha) [reply libraries/abstractions/ota_pal_psa/aws_ota_pal.c:347]: That is expected flow, thank you for sharing details.

File: libraries/abstractions/ota_pal_psa/aws_ota_pal.c (hunk 3 above)
Comment (pvyawaha) [reply libraries/abstractions/ota_pal_psa/aws_ota_pal.c:40]: The interface is slightly updated for OTA LTS 3.0 library and not for the pre-LTS library - https://github.com/aws/ota-for-aws-iot-embedded-sdk/blob/main/source/include/ota_platform_interface.h. So for pre-LTS OTA library, no changes are required for OTA PAL port.

**Comment by Sherryzhang2** [comment]
Any more comments on the PR?

**Pull request merged**